File: packages/generator-rust/src/bin/generate.rs
use std::io::{self, Read};

use crudui_generator::{
    bind_form, compile_form, render_form, render_list, AddRowOptions, BindOptions, CompileOptions,
    Form, FormError, FormResult, FormTemplate, ListOptions,
};
use serde::de::DeserializeOwned;
use serde_json::{json, Value};

fn input(message: impl Into<String>) -> FormError {
    FormError {
        code: "INVALID_FORM_INPUT".into(),
        message: message.into(),
        at: String::new(),
        trace: Vec::new(),
    }
}

fn decode<T: DeserializeOwned>(value: &Value) -> FormResult<T> {
    serde_json::from_value(value.clone()).map_err(|error| input(error.to_string()))
}

fn object<'a>(value: &'a Value, name: &str) -> FormResult<&'a Value> {
    if value.is_object() {
        Ok(value)
    } else {
        Err(input(format!("{name} must be an object")))
    }
}

fn option_string(options: &Value, key: &str) -> FormResult<Option<String>> {
    options
        .get(key)
        .map(|value| {
            value
                .as_str()
                .map(str::to_owned)
                .ok_or_else(|| input(format!("{key} must be a string")))
        })
        .transpose()
}

fn compile_options(options: &Value) -> FormResult<CompileOptions<'static>> {
    object(options, "options")?;
    Ok(CompileOptions {
        files: match options.get("files") {
            Some(files) => object(files, "files")?.as_object().unwrap().clone(),
            None => Default::default(),
        },
        basepath: option_string(options, "basepath")?.unwrap_or_default(),
        key_prefix: option_string(options, "keyPrefix")?,
        loader: None,
    })
}

fn error_json(error: FormError) -> Value {
    json!({"code":error.code,"message":error.message,"at":error.at})
}

fn string_arg(args: &[Value], index: usize) -> FormResult<&str> {
    args.get(index)
        .and_then(Value::as_str)
        .ok_or_else(|| input(format!("Argument {index} must be a string")))
}

fn action(form: &mut Form, action: &Value) -> FormResult<Value> {
    object(action, "action")?;
    let method = action["method"]
        .as_str()
        .ok_or_else(|| input("Action method must be a string"))?;
    let args = action["args"]
        .as_array()
        .ok_or_else(|| input("Action args must be an array"))?;
    match method {
        "setData" => {
            form.set_data(args.first().ok_or_else(|| input("setData requires data"))?)?;
        }
        "setValue" => {
            form.set_value(
                string_arg(args, 0)?,
                args.get(1)
                    .ok_or_else(|| input("setValue requires a value"))?
                    .clone(),
            )?;
        }
        "getData" => return Ok(form.get_data()),
        "getValue" => return form.get_value(string_arg(args, 0)?),
        "addRow" => {
            let options = args
                .get(1)
                .map(decode::<AddRowOptions>)
                .transpose()?
                .unwrap_or_default();
            return form
                .add_row(string_arg(args, 0)?, options)
                .map(Value::String);
        }
        "copyRow" => {
            let options = args
                .get(2)
                .map(decode::<AddRowOptions>)
                .transpose()?
                .unwrap_or_default();
            return form
                .copy_row(string_arg(args, 0)?, string_arg(args, 1)?, options)
                .map(Value::String);
        }
        "removeRow" => form.remove_row(string_arg(args, 0)?, string_arg(args, 1)?)?,
        "moveRow" => {
            let index = args
                .get(2)
                .and_then(Value::as_u64)
                .and_then(|index| usize::try_from(index).ok())
                .ok_or_else(|| input("Row position must be a nonnegative integer"))?;
            form.move_row(string_arg(args, 0)?, string_arg(args, 1)?, index)?;
        }
        "rekeyRow" => form.rekey_row(
            string_arg(args, 0)?,
            string_arg(args, 1)?,
            string_arg(args, 2)?,
        )?,
        _ => return Err(input(format!("Unknown form method: {method}"))),
    }
    Ok(Value::Null)
}

fn generate(request: &Value) -> FormResult<Value> {
    object(request, "request")?;
    let empty = json!({});
    let options = object(request.get("options").unwrap_or(&empty), "options")?;
    match request["operation"].as_str() {
        Some("compileForm") => {
            serde_json::to_value(compile_form(&request["spec"], &compile_options(options)?)?)
                .map_err(|error| input(error.to_string()))
        }
        Some("bindForm") => {
            let template: FormTemplate = decode(&request["template"])?;
            let data = object(request.get("data").unwrap_or(&empty), "data")?;
            Ok(bind_form(&template, data, &decode::<BindOptions>(options)?)?.into())
        }
        Some("renderList") => {
            let compilation = compile_options(options)?;
            let options = ListOptions {
                files: compilation.files,
                basepath: compilation.basepath,
                loader: None,
                language: option_string(options, "language")?.unwrap_or_else(|| "ko".into()),
                data: object(options.get("data").unwrap_or(&empty), "List context")?.clone(),
                page_meta: object(options.get("pageMeta").unwrap_or(&empty), "pageMeta")?
                    .as_object()
                    .unwrap()
                    .clone(),
                layout: option_string(options, "layout")?.unwrap_or_else(|| "table".into()),
            };
            let rows = match request.get("rows") {
                Some(rows) => rows
                    .as_array()
                    .ok_or_else(|| input("rows must be an array"))?
                    .as_slice(),
                None => &[],
            };
            render_list(&request["spec"], rows, &options).map(Value::String)
        }
        Some("form") => {
            let template: FormTemplate = decode(&request["template"])?;
            let data = object(request.get("data").unwrap_or(&empty), "data")?;
            let mut form = Form::new(template, data, decode(options)?)?;
            let mut steps = Vec::new();
            let actions = match request.get("actions") {
                Some(actions) => actions
                    .as_array()
                    .ok_or_else(|| input("actions must be an array"))?
                    .as_slice(),
                None => &[],
            };
            for step in actions {
                let (result, error) = match action(&mut form, step) {
                    Ok(value) => (value, Value::Null),
                    Err(error) => (Value::Null, error_json(error)),
                };
                steps.push(json!({"result":result,"error":error,"data":form.get_data(),"fields":form.fields(),"html":render_form(&form)?,"revision":form.revision()}));
            }
            Ok(
                json!({"data":form.get_data(),"fields":form.fields(),"html":render_form(&form)?,"revision":form.revision(),"steps":steps}),
            )
        }
        _ => Err(input("Unknown generator operation")),
    }
}

fn main() {
    let mut source = String::new();
    let result = io::stdin()
        .read_to_string(&mut source)
        .map_err(|error| input(error.to_string()))
        .and_then(|_| {
            let request: Value =
                serde_json::from_str(&source).map_err(|error| input(error.to_string()))?;
            generate(&request)
        });
    match result {
        Ok(value) => println!("{value}"),
        Err(error) => {
            println!("{}", json!({"error":error_json(error)}));
            std::process::exit(1);
        }
    }
}

File: tests/native-generators/programs/php/generate.php
<?php

declare(strict_types=1);

require __DIR__ . '/../../../../packages/generator-php/vendor/autoload.php';

use CRUDUI\Form;
use CRUDUI\FormError;
use CRUDUI\Generator;
use CRUDUI\Validator\Compose\ComposeLoadError;

/*
 * PHP generator process of the native generator conformance suite. It reads one JSON
 * request on standard input, calls the public API of crudui/generator and writes one
 * JSON value on standard output; ../../README.md defines the protocol. With the native
 * extension loaded, the same program runs the extension's classes.
 */

function failure(Throwable $error): array
{
    return ['code' => $error instanceof ComposeLoadError || $error instanceof FormError ? $error->getErrorCode() : 'INVALID_FORM_INPUT', 'message' => $error->getMessage(), 'at' => $error instanceof FormError ? $error->getPath() : ($error instanceof ComposeLoadError ? implode('.', $error->getCompositionTrace()) : '')];
}
function state(Form $form): array
{
    return ['data' => $form->getData(), 'fields' => $form->getFields(), 'html' => Generator::renderForm($form), 'revision' => $form->getRevision()];
}
/** A protocol JSON object; a JSON array or scalar is not an object in any runtime. */
function objectValue(mixed $value, string $message): stdClass
{
    if (!$value instanceof stdClass) {
        throw new InvalidArgumentException($message);
    }
    return $value;
}
/** The request member, or null when it is absent. */
function member(stdClass $request, string $key): mixed
{
    return property_exists($request, $key) ? $request->{$key} : null;
}
/** A form template is a JSON object; the library checks its kind. */
function template(stdClass $request): stdClass
{
    return objectValue(member($request, 'template'), 'Unsupported form template');
}
function options(mixed $value): array
{
    if (!$value instanceof stdClass) {
        throw new InvalidArgumentException('Options must be an object');
    }
    return (array) $value;
}
try {
    try {
        $request = json_decode(stream_get_contents(STDIN), false, 512, JSON_THROW_ON_ERROR);
    } catch (JsonException) {
        throw new InvalidArgumentException('Request must be valid JSON');
    }
    if (!$request instanceof stdClass) {
        throw new InvalidArgumentException('Request must be an object');
    }
    switch ($request->operation ?? null) {
        case 'compileForm':
            $result = Generator::compileForm(objectValue(member($request, 'spec'), 'A form spec must be a group with properties'), options(property_exists($request, 'options') ? $request->options : new stdClass()));
            break;
        case 'bindForm':
            $result = Generator::bindForm(template($request), property_exists($request, 'data') ? objectValue($request->data, 'Form data must be an object') : new stdClass(), options(property_exists($request, 'options') ? $request->options : new stdClass()));
            break;
        case 'bindButtons':
            $result = Generator::bindButtons(template($request), property_exists($request, 'data') ? objectValue($request->data, 'Form data must be an object') : new stdClass(), options(property_exists($request, 'options') ? $request->options : new stdClass()));
            break;
        case 'formButtonsHtml':
            $buttons = member($request, 'buttons');
            if (!is_array($buttons)) {
                throw new InvalidArgumentException('Form buttons must be a list');
            }
            $result = Generator::formButtonsHtml($buttons);
            break;
        case 'renderList':
            // JSON decides each type before the library applies the PHP value rules.
            $spec = objectValue(member($request, 'spec'), 'List specification must be an object');
            $rows = property_exists($request, 'rows') ? $request->rows : [];
            if (!is_array($rows)) {
                throw new InvalidArgumentException('List rows must be an array');
            }
            foreach ($rows as $row) {
                objectValue($row, 'List rows must be objects');
            }
            $listOptions = options(property_exists($request, 'options') ? $request->options : new stdClass());
            if (isset($listOptions['data'])) {
                objectValue($listOptions['data'], 'List context must be an object');
            }
            $result = Generator::renderList($spec, $rows, $listOptions);
            break;
        case 'buildList':
            $spec = objectValue(member($request, 'spec'), 'List specification must be an object');
            $rows = property_exists($request, 'rows') ? $request->rows : [];
            if (!is_array($rows)) {
                throw new InvalidArgumentException('List rows must be an array');
            }
            foreach ($rows as $row) {
                objectValue($row, 'List rows must be objects');
            }
            $listOptions = options(property_exists($request, 'options') ? $request->options : new stdClass());
            if (isset($listOptions['data'])) {
                objectValue($listOptions['data'], 'List context must be an object');
            }
            $result = Generator::buildList($spec, $rows, $listOptions);
            break;
        case 'buildDetail':
        case 'renderDetail':
            $method = $request->operation;
            $spec = objectValue(member($request, 'spec'), 'Detail specification must be an object');
            $record = property_exists($request, 'record') ? objectValue($request->record, 'Detail record must be an object') : new stdClass();
            $detailOptions = options(property_exists($request, 'options') ? $request->options : new stdClass());
            // The context is checked after the fields, which the library checks first.
            if (property_exists($spec, 'fields') && isset($detailOptions['data'])) {
                objectValue($detailOptions['data'], 'Detail context must be an object');
            }
            $result = Generator::$method($spec, $record, $detailOptions);
            break;
        case 'form':
            $form = new Form(template($request), property_exists($request, 'data') ? objectValue($request->data, 'Form data must be an object') : new stdClass(), options(property_exists($request, 'options') ? $request->options : new stdClass()));
            $steps = [];
            $actions = property_exists($request, 'actions') ? $request->actions : [];
            if (!is_array($actions)) {
                throw new InvalidArgumentException('Actions must be an array');
            }
            foreach ($actions as $action) {
                $value = null;
                $error = null;
                try {
                    $method = $action instanceof stdClass ? ($action->method ?? null) : null;
                    $args = $action instanceof stdClass && property_exists($action, 'args') ? $action->args : null;
                    if (!in_array($method, ['setData', 'setValue', 'addRow', 'copyRow', 'removeRow', 'moveRow', 'rekeyRow', 'getValue', 'getData'], true) || !is_array($args)) {
                        throw new InvalidArgumentException('Invalid form action');
                    }
                    if ($method === 'setData' && array_key_exists(0, $args)) {
                        $args[0] = objectValue($args[0], 'Form data must be an object');
                    }
                    $optionIndex = $method === 'addRow' ? 1 : ($method === 'copyRow' ? 2 : null);
                    if ($optionIndex !== null && array_key_exists($optionIndex, $args)) {
                        $args[$optionIndex] = options($args[$optionIndex]);
                    }
                    $value = $form->{$method}(...$args);
                } catch (Throwable $caught) {
                    $error = failure($caught);
                }
                $steps[] = ['result' => $value, 'error' => $error, ...state($form)];
            }
            $result = [...state($form), 'steps' => $steps];
            break;
        default:
            throw new InvalidArgumentException('Unknown generator operation');
    }
    echo json_encode($result, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_THROW_ON_ERROR) . "\n";
} catch (Throwable $error) {
    echo json_encode(['error' => failure($error)], JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_THROW_ON_ERROR) . "\n";
    exit(1);
}
